src/darwin/Framework/CHIP/MTRBaseDevice.mm

```diff
@@ -979,32 +1005,25 @@ - (void)readAttributePaths:(NSArray<MTRAttributeRequestPath *> * _Nullable)attri
                 }
             };
 
-            AttributePathParams * attributePathParamsList = nullptr;
-            EventPathParams * eventPathParamsList = nullptr;
+            Platform::MemoryUniquePtr<AttributePathParams> attributePathParamsList = nullptr;
```

Comment:
Now that I am looking at this again... was there a reason we couldn't use the existing `ScopedMemoryBuffer` class here?  It does have a `Release()`, method, so seems like it should be usable here.

So to be precise, here is what I am thinking this should look like:

```
   __block AttributePathParams * attributePathParamsToFree = nullptr;
  ScopedMemoryBuffer<AttributePathParams> attributePathParamsList;
  VerifyOrReturnError(attributePathParamsList.Calloc([attributes count], CHIP_ERROR_NO_MEMORY);
  ...
  // Fill in the values, do the same for the events, kick off the read.  If he read kicks off successfully:
  attributePathParamsToFree = attributePathParamsList.Release();
```

and in `onDone` you free attributePathParamsToFree if it's not null.  No changes needed to the callback compared to what you had before, or to the ReadClientContainer.

Reply:
I understand your intention, but it is difficult to implement with your method because the value of "attributePathParamsToFree" is not determined at the time of onDone's lambda definition code.

I'll think about it a bit more and add additional comments.

Reply:
I implemented it using ScopeBuffer.
However, since it is captured using Get() in advance, there is no problem in implementation, but there is a problem that even if it is not null, it cannot guarantee whether it is Free or not.

(I think there is no problem in the code because onDone is not called if an error occurs in the callback or readClient after onDone is declared.)

Reply:
Oh, right, onDone is a C++ lambda, not an Objective C block...

But yes, what you ended up doing should work, exactly because onDone will not be called unless we succeed at kicking off the read.  Good catch on that.